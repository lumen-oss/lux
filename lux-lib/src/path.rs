use itertools::Itertools;
use path_slash::PathBufExt;
use serde::Serialize;
use std::{env, fmt::Display, path::PathBuf, str::FromStr};
use thiserror::Error;

use crate::{
    build::utils::c_dylib_extension,
    config::{Config, LuaVersion},
    tree::{Tree, TreeError},
};

const LUA_PATH_SEPARATOR: &str = ";";
const LUA_INIT: &str = "require('lux').loader()";

#[derive(PartialEq, Eq, Debug, Serialize)]
pub struct Paths {
    /// Paths for Lua libraries
    src: PackagePath,
    /// Paths for native Lua libraries
    lib: PackagePath,
    /// Paths for executables
    bin: BinPath,

    version: LuaVersion,
}

#[derive(Debug, Error)]
pub enum PathsError {
    #[error(transparent)]
    Tree(#[from] TreeError),
}

impl Paths {
    fn default(tree: &Tree) -> Self {
        Self {
            src: <_>::default(),
            lib: <_>::default(),
            bin: <_>::default(),
            version: tree.version().clone(),
        }
    }

    pub fn new(tree: &Tree) -> Result<Self, PathsError> {
        let mut paths = tree
            .list()?
            .into_iter()
            .flat_map(|(_, packages)| {
                packages
                    .into_iter()
                    .map(|package| tree.installed_rock_layout(&package))
                    .collect_vec()
            })
            .try_fold(Self::default(tree), |mut paths, package| {
                let package = package?;
                paths.src.0.push(package.src.join("?.lua"));
                paths.src.0.push(package.src.join("?").join("init.lua"));
                paths
                    .lib
                    .0
                    .push(package.lib.join(format!("?.{}", c_dylib_extension())));
                paths.bin.0.push(package.bin);
                Ok::<Paths, TreeError>(paths)
            })?;

        let lib_name = format!("lux-lua{}", tree.version());
        let lib_path = option_env!("LUX_LIB_DIR")
            .map(PathBuf::from)
            .unwrap_or_else(|| {
                pkg_config::Config::new()
                    .print_system_libs(false)
                    .cargo_metadata(false)
                    .env_metadata(false)
                    .probe(&lib_name)
                    .ok()
                    .and_then(|library| library.link_paths.first().cloned())
                    .unwrap_or_else(|| Config::get_default_data_path().unwrap())
            })
            .join(tree.version().to_string())
            .join("?.so");

        paths.prepend(&Paths {
            version: tree.version().clone(),
            src: <_>::default(),
            bin: <_>::default(),
            lib: PackagePath(vec![lib_path]),
        });

        Ok(paths)
    }

    /// Get the `package.path`
    pub fn package_path(&self) -> &PackagePath {
        &self.src
    }

    /// Get the `package.cpath`
    pub fn package_cpath(&self) -> &PackagePath {
        &self.lib
    }

    /// Get the `$PATH`
    pub fn path(&self) -> &BinPath {
        &self.bin
    }

    /// Get `$LUA_INIT`
    pub fn init(&self) -> String {
        format!("if _VERSION:find('{}') then {LUA_INIT} end", self.version)
    }

    /// Get the `$PATH`, prepended to the existing `$PATH` environment.
    pub fn path_prepended(&self) -> BinPath {
        let mut path = BinPath::from_env();
        path.prepend(self.path());
        path
    }

    pub fn prepend(&mut self, other: &Self) {
        self.src.prepend(&other.src);
        self.lib.prepend(&other.lib);
        self.bin.prepend(&other.bin);
    }
}

#[derive(PartialEq, Eq, Debug, Default, Serialize)]
pub struct PackagePath(Vec<PathBuf>);

impl PackagePath {
    pub fn prepend(&mut self, other: &Self) {
        let mut new_vec = other.0.to_owned();
        new_vec.append(&mut self.0);
        self.0 = new_vec;
    }
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
    pub fn joined(&self) -> String {
        self.0
            .iter()
            .unique()
            .map(|path| path.to_slash_lossy())
            .join(LUA_PATH_SEPARATOR)
    }
}

impl FromStr for PackagePath {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let paths = s
            .trim_start_matches(LUA_PATH_SEPARATOR)
            .trim_end_matches(LUA_PATH_SEPARATOR)
            .split(LUA_PATH_SEPARATOR)
            .map(PathBuf::from)
            .collect();
        Ok(PackagePath(paths))
    }
}

impl Display for PackagePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.joined().fmt(f)
    }
}

#[derive(PartialEq, Eq, Debug, Default, Serialize)]
pub struct BinPath(Vec<PathBuf>);

impl BinPath {
    pub fn from_env() -> Self {
        Self::from_str(env::var("PATH").unwrap_or_default().as_str()).unwrap_or_default()
    }
    pub fn prepend(&mut self, other: &Self) {
        let mut new_vec = other.0.to_owned();
        new_vec.append(&mut self.0);
        self.0 = new_vec;
    }
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
    pub fn joined(&self) -> String {
        env::join_paths(self.0.iter().unique())
            .expect("Failed to join bin paths.")
            .to_string_lossy()
            .to_string()
    }
}

impl FromStr for BinPath {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let paths = env::split_paths(s).collect();
        Ok(BinPath(paths))
    }
}

impl Display for BinPath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.joined().fmt(f)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn package_path_leading_trailing_delimiters() {
        let path = PackagePath::from_str(
            ";;/path/to/some/lib/lua/5.1/?.so;/path/to/another/lib/lua/5.1/?.so;;;",
        )
        .unwrap();
        assert_eq!(
            path,
            PackagePath(vec![
                "/path/to/some/lib/lua/5.1/?.so".into(),
                "/path/to/another/lib/lua/5.1/?.so".into(),
            ])
        );
        assert_eq!(
            format!("{}", path),
            "/path/to/some/lib/lua/5.1/?.so;/path/to/another/lib/lua/5.1/?.so"
        );
    }
}
