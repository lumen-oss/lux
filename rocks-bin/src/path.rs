use std::{env, str::FromStr as _};

use clap::Subcommand;
use eyre::Result;
use rocks_lib::{
    config::{Config, LuaVersion},
    path::{BinPath, PackagePath, Paths},
    tree::Tree,
};
use strum::{EnumString, VariantNames};
use strum_macros::Display;

use clap::{Args, ValueEnum};

#[derive(Args)]
pub struct Path {
    #[command(subcommand)]
    cmd: Option<PathCmd>,

    /// Append the rocks tree paths to the system paths.
    #[clap(default_value_t = false)]
    #[arg(long)]
    append: bool,
}

#[derive(Subcommand, PartialEq, Eq, Debug, Clone)]
#[clap(rename_all = "kebab_case")]
enum PathCmd {
    /// Generate an export statement for all paths.
    /// (formatted as a shell command) [Default]
    Full(FullArgs),
    /// Generate a `LUA_PATH` expression for `lua` libraries in the rocks tree.
    /// (not formatted as a shell command)
    Lua,
    /// Generate a `LUA_CPATH` expression for native `lib` libraries in the rocks tree.
    /// (not formatted as a shell command)
    C,
    /// Generate a `PATH` expression for `bin` executables in the rocks tree.
    /// (not formatted as a shell command)
    Bin,
}

impl Default for PathCmd {
    fn default() -> Self {
        Self::Full(FullArgs::default())
    }
}

#[derive(Args, PartialEq, Eq, Debug, Clone, Default)]
struct FullArgs {
    /// Do not export `PATH` (`bin` paths).
    #[clap(default_value_t = false)]
    #[arg(long)]
    no_bin: bool,

    /// The shell to format for.
    #[clap(default_value_t = Shell::default())]
    #[arg(long)]
    shell: Shell,
}

#[derive(EnumString, VariantNames, Display, ValueEnum, PartialEq, Eq, Debug, Clone)]
#[strum(serialize_all = "lowercase")]
enum Shell {
    Posix,
    Fish,
    Nu,
}

impl Default for Shell {
    fn default() -> Self {
        Self::Posix
    }
}

pub async fn path(path_data: Path, config: Config) -> Result<()> {
    let tree = Tree::new(config.tree().clone(), LuaVersion::from(&config)?)?;
    let paths = Paths::from_tree(tree)?;
    let cmd = path_data.cmd.unwrap_or_default();
    let append = path_data.append;
    match cmd {
        PathCmd::Full(args) => {
            let mut result = String::new();
            let shell = args.shell;
            let package_path = mk_package_path(&paths, append)?;
            if !package_path.is_empty() {
                result.push_str(format_export(&shell, "LUA_PATH", &package_path).as_str());
                result.push('\n')
            }
            let package_cpath = mk_package_cpath(&paths, append)?;
            if !package_cpath.is_empty() {
                result.push_str(format_export(&shell, "LUA_CPATH", &package_cpath).as_str());
                result.push('\n')
            }
            if !args.no_bin {
                let path = mk_bin_path(&paths, append)?;
                if !path.is_empty() {
                    result.push_str(format_export(&shell, "PATH", &path).as_str());
                    result.push('\n')
                }
            }
            println!("{}", &result);
        }
        PathCmd::Lua => println!("{}", &mk_package_path(&paths, append)?),
        PathCmd::C => println!("{}", &mk_package_cpath(&paths, append)?),
        PathCmd::Bin => println!("{}", &mk_bin_path(&paths, append)?),
    }
    Ok(())
}

fn mk_package_path(paths: &Paths, append: bool) -> Result<PackagePath> {
    let mut result = if append {
        PackagePath::from_str(env::var("LUA_PATH").unwrap_or_default().as_str()).unwrap_or_default()
    } else {
        PackagePath::default()
    };
    result.append(paths.package_path());
    Ok(result)
}

fn mk_package_cpath(paths: &Paths, append: bool) -> Result<PackagePath> {
    let mut result = if append {
        PackagePath::from_str(env::var("LUA_CPATH").unwrap_or_default().as_str())
            .unwrap_or_default()
    } else {
        PackagePath::default()
    };
    result.append(paths.package_cpath());
    Ok(result)
}

fn mk_bin_path(paths: &Paths, append: bool) -> Result<BinPath> {
    let mut result = if append {
        BinPath::from_env()
    } else {
        BinPath::default()
    };
    result.append(paths.path());
    Ok(result)
}

fn format_export<D>(shell: &Shell, var_name: &str, var: &D) -> String
where
    D: std::fmt::Display,
{
    match shell {
        Shell::Posix => format!("export {}='{}';", var_name, var),
        Shell::Fish => format!("set -x {} \"{}\";", var_name, var),
        Shell::Nu => format!("$env.{} = \"{}\";", var_name, var),
    }
}
