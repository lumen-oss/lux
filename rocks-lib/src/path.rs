use itertools::Itertools;
use serde::Serialize;
use std::{env, fmt::Display, io, path::PathBuf, str::FromStr};

use crate::{rockspec::utils::lua_lib_extension, tree::Tree};

const LUA_PATH_SEPARATOR: &str = ";";

#[derive(PartialEq, Eq, Debug, Default, Serialize)]
pub struct Paths {
    /// Paths for Lua libraries
    src: PackagePath,
    /// Paths for native Lua libraries
    lib: PackagePath,
    /// Paths for executables
    bin: BinPath,
}

impl Paths {
    pub fn from_tree(tree: Tree) -> io::Result<Self> {
        let paths = tree
            .list()?
            .into_iter()
            .flat_map(|(_, packages)| {
                packages
                    .into_iter()
                    .map(|package| tree.rock_layout(&package))
                    .collect_vec()
            })
            .fold(Self::default(), |mut paths, package| {
                paths.src.0.push(package.src.join("?.lua"));
                paths.src.0.push(package.src.join("?").join("init.lua"));
                paths
                    .lib
                    .0
                    .push(package.lib.join(format!("?.{}", lua_lib_extension())));
                paths.bin.0.push(package.bin);
                paths
            });
        Ok(paths)
    }

    /// Get the `package.path`
    pub fn package_path(&self) -> &PackagePath {
        &self.src
    }

    /// Get the `package.cpath`
    pub fn package_cpath(&self) -> &PackagePath {
        &self.lib
    }

    /// Get the `$PATH`
    pub fn path(&self) -> &BinPath {
        &self.bin
    }

    /// Get the `$PATH`, appended to the existing `$PATH` environment.
    pub fn path_appended(&self) -> BinPath {
        let mut path = BinPath::from_env();
        path.append(self.path());
        path
    }
}

#[derive(PartialEq, Eq, Debug, Default, Serialize)]
pub struct PackagePath(Vec<PathBuf>);

impl PackagePath {
    pub fn append(&mut self, other: &Self) {
        self.0.extend(other.0.to_owned())
    }
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
    pub fn joined(&self) -> String {
        self.0
            .iter()
            .map(|path| path.to_string_lossy())
            .join(LUA_PATH_SEPARATOR)
    }
}

impl FromStr for PackagePath {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let paths = s
            .trim_start_matches(LUA_PATH_SEPARATOR)
            .trim_end_matches(LUA_PATH_SEPARATOR)
            .split(LUA_PATH_SEPARATOR)
            .map(PathBuf::from)
            .collect();
        Ok(PackagePath(paths))
    }
}

impl Display for PackagePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.joined().fmt(f)
    }
}

#[derive(PartialEq, Eq, Debug, Default, Serialize)]
pub struct BinPath(Vec<PathBuf>);

impl BinPath {
    pub fn from_env() -> Self {
        Self::from_str(env::var("PATH").unwrap_or_default().as_str()).unwrap_or_default()
    }
    pub fn append(&mut self, other: &Self) {
        self.0.extend(other.0.to_owned())
    }
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
    pub fn joined(&self) -> String {
        env::join_paths(self.0.iter())
            .expect("Failed to join bin paths.")
            .to_string_lossy()
            .to_string()
    }
}

impl FromStr for BinPath {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let paths = env::split_paths(s).collect();
        Ok(BinPath(paths))
    }
}

impl Display for BinPath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.joined().fmt(f)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn package_path_leading_trailing_delimiters() {
        let path = PackagePath::from_str(
            ";;/path/to/some/lib/lua/5.1/?.so;/path/to/another/lib/lua/5.1/?.so;;;",
        )
        .unwrap();
        assert_eq!(
            path,
            PackagePath(vec![
                "/path/to/some/lib/lua/5.1/?.so".into(),
                "/path/to/another/lib/lua/5.1/?.so".into(),
            ])
        );
        assert_eq!(
            format!("{}", path),
            "/path/to/some/lib/lua/5.1/?.so;/path/to/another/lib/lua/5.1/?.so"
        );
    }
}
