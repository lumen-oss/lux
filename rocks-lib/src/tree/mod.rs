use std::path::PathBuf;

use crate::{
    config::LuaVersion,
    lua_package::{LuaPackage, LuaPackageReq, PackageName, PackageVersion},
};
use eyre::Result;

mod list;

/// A tree is a collection of files where installed rocks are located.
///
/// `rocks` diverges from the traditional hierarchy employed by luarocks.
/// Instead, we opt for a much simpler approach:
///
/// - /rocks/<lua-version> - contains rocks
/// - /rocks/<lua-version>/<rock>/etc - documentation and supplementary files for the rock
/// - /rocks/<lua-version>/<rock>/lib - shared libraries (.so files)
/// - /rocks/<lua-version>/<rock>/src - library code for the rock
/// - /bin - binary files produced by various rocks

#[derive(Clone, Debug)]
pub struct Tree {
    /// The Lua version of the tree.
    version: LuaVersion,
    /// The root of the tree.
    root: PathBuf,
}

/// Change-agnostic way of referencing various paths for a rock.
#[derive(Debug, PartialEq)]
pub struct RockLayout {
    pub etc: PathBuf,
    pub lib: PathBuf,
    pub src: PathBuf,
}

impl Tree {
    pub fn new(root: PathBuf, version: LuaVersion) -> Result<Self> {
        // Ensure that the root and the version directory exist.
        std::fs::create_dir_all(root.join(version.to_string()))?;

        // Ensure that the bin directory exists.
        std::fs::create_dir_all(root.join("bin"))?;

        Ok(Self { root, version })
    }

    pub fn root(&self) -> PathBuf {
        self.root.join(self.version.to_string())
    }

    pub fn root_for(&self, rock_name: &PackageName, rock_version: &PackageVersion) -> PathBuf {
        self.root().join(format!("{}@{}", rock_name, rock_version))
    }

    pub fn bin(&self) -> PathBuf {
        self.root.join("bin")
    }

    pub fn has_rock(&self, req: &LuaPackageReq) -> Option<LuaPackage> {
        self.list().get(req.name()).map(|versions| {
            versions.iter().rev().find_map(|version| {
                if req.version_req().matches(version) {
                    Some(LuaPackage::new(req.name().clone(), version.clone()))
                } else {
                    None
                }
            })
        })?
    }

    pub fn rock(
        &self,
        rock_name: &PackageName,
        rock_version: &PackageVersion,
    ) -> Result<RockLayout> {
        // TODO(vhyrro): Don't store rocks with the revision number, that should be stripped almost
        // everywhere by default.
        let rock_path = self.root_for(rock_name, rock_version);

        let etc = rock_path.join("etc");
        let lib = rock_path.join("lib");
        let src = rock_path.join("src");

        std::fs::create_dir_all(&etc)?;
        std::fs::create_dir_all(&lib)?;
        std::fs::create_dir_all(&src)?;

        Ok(RockLayout { etc, lib, src })
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::{assert_yaml_snapshot, sorted_redaction};

    use crate::{config::LuaVersion, tree::RockLayout};

    use super::Tree;

    #[test]
    fn rock_layout() {
        let tree_path =
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("resources/test/sample-tree");

        let tree = Tree::new(tree_path.clone(), LuaVersion::Lua51).unwrap();

        let neorg = tree
            .rock(&"neorg".into(), &"8.0.0".parse().unwrap())
            .unwrap();

        assert_eq!(
            neorg,
            RockLayout {
                etc: tree_path.join("5.1/neorg@8.0.0/etc"),
                lib: tree_path.join("5.1/neorg@8.0.0/lib"),
                src: tree_path.join("5.1/neorg@8.0.0/src"),
            }
        );

        let lua_cjson = tree
            .rock(&"lua-cjson".into(), &"2.1.0".parse().unwrap())
            .unwrap();

        assert_eq!(
            lua_cjson,
            RockLayout {
                etc: tree_path.join("5.1/lua-cjson@2.1.0/etc"),
                lib: tree_path.join("5.1/lua-cjson@2.1.0/lib"),
                src: tree_path.join("5.1/lua-cjson@2.1.0/src"),
            }
        );
    }

    #[test]
    fn tree_list() {
        let tree_path =
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("resources/test/sample-tree");

        let tree = Tree::new(tree_path, LuaVersion::Lua51).unwrap();

        assert_yaml_snapshot!(tree.list(), { "." => sorted_redaction() })
    }
}
