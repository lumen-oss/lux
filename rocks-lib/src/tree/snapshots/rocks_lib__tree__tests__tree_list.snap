---
source: rocks-lib/src/tree/mod.rs
assertion_line: 252
expression: sorted_result
---
- - lua-cjson
  - - 2.1.0-1
- - neorg
  - - 8.0.0-1
    - 8.8.1-1
